### PR TITLE
[SPARK-38529][SQL] Prevent GeneratorNestedColumnAliasing to be applied to non-Explode generators

### PR DESCRIPTION
**Pull request by minyyy**
### What changes were proposed in this pull request?
1. Explicitly return in GeneratorNestedColumnAliasing when the generator is not Explode.
2. Add extensive comment to GeneratorNestedColumnAliasing.
3. An off-hand code refactor to make the code clearer.

### Why are the changes needed?
GeneratorNestedColumnAliasing does not handle other generators correctly. We only try to rewrite the generator for Explode but try to rewrite all ExtractValue expressions. This can cause inconsistency for non-Explode generators.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 1]
```diff
@@ -395,12 +434,11 @@ object GeneratorNestedColumnAliasing {
             val updatedGenerate = rewrittenG.copy(generatorOutput = updatedGeneratorOutput)
 
             // Replace nested column accessor with generator output.
-            val attrExprIdsOnGenerator = attrToExtractValuesOnGenerator.keys.map(_.exprId).toSet
+            val generateOutputAttrExprId = attrToExtractValuesOnGenerator.keys.map(_.exprId).head
```
Comment (minyyy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:437]: I changed the toSet and find way to toSet.head to make it clearer - we should only have one generator output here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 2]
```diff
@@ -395,12 +434,11 @@ object GeneratorNestedColumnAliasing {
...
+            val generateOutputAttr = updatedGenerate.output
+              .find(_.exprId == generateOutputAttrExprId).get
```
Comment (minyyy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:439]: I changed the .getOrElse to .get to make it clearer - we should never fail to find the generator output attribute, if we want to be safer, we can use a require to check and return here, but should not use getOrElse because IMO if we run into the "else" branch it means an error.

**Comment by HyukjinKwon** [comment]
cc @viirya FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 3]
```diff
@@ -348,6 +382,11 @@ object GeneratorNestedColumnAliasing {
         case _ =>
       }
 
+      g.generator match {
+        case _: ExplodeBase =>
+        case _ => return Some(pushedThrough)
+      }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:388]: I think you can just remove `Inline` from `canPruneGenerator`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 4]
```diff
@@ -321,6 +321,40 @@ object GeneratorNestedColumnAliasing {
     // need to prune nested columns through Project and under Generate. The difference is
     // when `nestedSchemaPruningEnabled` is on, nested columns will be pruned further at
     // file format readers if it is supported.
+
+    // There are [[ExtractValue]] expressions on or not on the output of the generator. Generator
+    // can also have different types:
+    // 1. For [[ExtractValue]]s not on the output of the generator, theoretically speaking, there
+    //    lots of expressions that we can push down, including non ExtractValues and GetArrayItem
+    //    and GetMapValue. But to be safe, we only handle GetStructField and GetArrayStructFields.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:329]: Not sure if this is correct. We don't actually limit the `ExtractValue` types.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 5]
```diff
@@ -321,6 +321,40 @@ object GeneratorNestedColumnAliasing {
...
+    // 2. For [[ExtractValue]]s on the output of the generator, the situation depends on the type
+    //    of the generator expression.
+    //   2.1 Inline
```
[the remark this answers is not in the document]
Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:332]: The reason that I don't remove `Inline` from `canPruneGenerator` is that I don't want to change our existing behavior of the pushdown of exprs not on generators, if I change `canPruneGenerator` then less expressions enter this branch, we call NestedColumnAliasing for less expressions. But if you are fine with it, I can simply change to it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 3 above)
Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:388]: Discussed below.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 6]
```diff
@@ -321,6 +321,40 @@ object GeneratorNestedColumnAliasing {
...
+    //       GetArrayStructFields, since the pushed down expression will operate on an array of
+    //       array which is invalid.
+    //   2.3 Stack
+    //       Stack takes a sequence of expressions, and returns an output of
+    //       STRUCT<col0: some_type, col1: some_type, ...>
+    //       The push down is doable but more complicated in this case as the expression that
+    //       operates on the col_i of the output needs to pushed down to every (kn+i)-th input
+    //       expression where n is the total number of columns (or struct fields) of the output.
+
+    // To be safe, we only support
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:357]: Looks like a broken sentence.

**Review by viirya — COMMENTED** [review]
Actually the optimizer rule only transforms `ExplodeBase`, although `canPruneGenerator` lists `Stack` and `Inline`. You can just remove `Stack` and `Inline` from `canPruneGenerator`, I think.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 4 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:329]: We have a check in the rule. For certain generator input, e.g. Map, they will not be handled now.

Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:329]: I will rephrase it - We do not support them for now. I agree that if we want we can support it if we add some expressions like `GetArrayMapValue`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:332]: I think it's okay to do. As they are not actually handled now, excluding them won't change the behavior and can prevent possible bugs in the future. When we can handle them in the rule, we can add them back explicitly.

**Comment by minyyy** [comment]
Changed `canPruneGenerator` per the comment.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala [hunk 7]
```diff
@@ -812,6 +812,18 @@ class NestedColumnAliasingSuite extends SchemaPruningTest {
     val expected3 = contact.select($"name").rebalance($"name").select($"name.first").analyze
     comparePlans(optimized3, expected3)
   }
+
+  test("GeneratorNestedColumnAliasing does not pushdown for non-Explode") {
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala:816]: Can you add JIRA ticket as prefix? E.g. `SPARK-38529: GeneratorNestedColumnAliasing does not pushdown for non-Explode`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala [hunk 8]
```diff
@@ -292,7 +292,7 @@ class NestedColumnAliasingSuite extends SchemaPruningTest {
     comparePlans(optimized, expected)
   }
 
-  test("Nested field pruning for Project and Generate") {
+  test("Nested field pruning for Project and Explode") {
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala:295]: This is unnecessary change. Can you revert it back?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 9]
```diff
@@ -447,9 +478,6 @@ object GeneratorNestedColumnAliasing {
    */
   def canPruneGenerator(g: Generator): Boolean = g match {
     case _: Explode => true
-    case _: Stack => true
-    case _: PosExplode => true
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:451]: Do we need to remove `PosExplode`?

Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:451]: reverted.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala (hunk 8 above)
Comment (minyyy) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala]: Done.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala (hunk 7 above)
Comment (minyyy) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala]: Done.

**Review by viirya — DISMISSED** [review]

**Comment by viirya** [comment]
BTW, I will only merge this to master and not backport it. It is because previously user queries are not failed. I will avoid backpointing this to prevent possible regression. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 10]
```diff
@@ -321,6 +321,38 @@ object GeneratorNestedColumnAliasing {
     // need to prune nested columns through Project and under Generate. The difference is
     // when `nestedSchemaPruningEnabled` is on, nested columns will be pruned further at
     // file format readers if it is supported.
+
+    // There are [[ExtractValue]] expressions on or not on the output of the generator. Generator
+    // can also have different types:
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:326]: Can you rephrase the comment? It looks a bit weird. E.g., I'm not sure what the first sentence means/wants to mean.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 11]
```diff
@@ -321,6 +321,38 @@ object GeneratorNestedColumnAliasing {
...
+    // 1. For [[ExtractValue]]s not on the output of the generator, theoretically speaking, there
+    //    lots of expressions that we can push down, including non ExtractValues and GetArrayItem
+    //    and GetMapValue. But to be safe, we only handle GetStructField and GetArrayStructFields.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:329]: The first item looks a bit weird. For what `ExtractValue` can be pushed down, you can simply list them. I'm not sure "For [[ExtractValue]]s not on the output of the generator..." means. Do you mean on the output of generator?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 12]
```diff
@@ -321,6 +321,38 @@ object GeneratorNestedColumnAliasing {
...
+    //       For the array case, it is only valid to push down GetStructField. After push down,
+    //       the GetStructField becomes a GetArrayStructFields. Note that we cannot push down
+    //       GetArrayStructFields, since the pushed down expression will operate on an array of
+    //       array which is invalid.
+    //   2.3 Stack
+    //       Stack takes a sequence of expressions, and returns an output of
+    //       STRUCT<col0: some_type, col1: some_type, ...>
+    //       The push down is doable but more complicated in this case as the expression that
+    //       operates on the col_i of the output needs to pushed down to every (kn+i)-th input
+    //       expression where n is the total number of columns (or struct fields) of the output.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:355]: For Stack and Inline, it looks verbose and doesn't help on reading the code. Can you just mention we only support Explode.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 11 above)
Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:329]:
```
val (attrToExtractValuesOnGenerator, attrToExtractValuesNotOnGenerator) =
  attrToExtractValues.partition { case (attr, _) =>
    attr.references.subsetOf(generatorOutputSet) }

val pushedThrough = NestedColumnAliasing.rewritePlanWithAliases(
  plan, attrToExtractValuesNotOnGenerator)
```

We call `NestedColumnAliasing.rewritePlanWithAliases` for `ExtractValue`s not on the output of the generators. The above code snippet is also the reason that I was hesitated to modify `canPruneGenerator`. If we modify `canPruneGenerator`, we miss the opportunity for `Inline` and `Stack` to use `NestedColumnAliasing.rewritePlanWithAliases` on `attrToExtractValuesNotOnGenerator`. My point in the comment was to note the difference of optimizations able to apply to `attrToExtractValuesOnGenerator` and `attrToExtractValuesNotOnGenerator`. The former one is predicate pushdown and there are lots of expressions that can be pushed down, the latter one is the real "GeneratorNestedColumnAliasing" and only `GetStructField` is supported.

**Comment by minyyy** [comment]
@cloud-fan Could you help with reviewing this PR please? Thanks.

**Comment by minyyy** [comment]
Changed back to my previous implementation - Now I still check `.isInstanceOf[ExplodeBase]` in the case match branch instead of modifying `canPruneGenerator` since we do have tests covering the pushdown of fields not on the output of the generator list (the `Nested column pruning for Generate` test in `ColumnPruningSuite.scala`).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 12 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:376]: actually, I find it's useful to understand why we only support explode today.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3!
